### PR TITLE
add FSDP QLoRA test and revert failing PR

### PR DESCRIPTION
**Pull request by weifengpy**
fix error when running torchtune QLoRA + FSDP2 https://github.com/pytorch/ao/issues/380
`TypeError: nf4_detach() missing 1 required positional argument: 'args'`

torchtune command
```
tune download meta-llama/Llama-2-7b-hf --output-dir /tmp/Llama-2-7b-hf --hf-token <HF_TOKEN>
tune run --nnodes 1 --nproc_per_node 4 lora_finetune_fsdp2 --config llama2/7B_qlora enable_activation_checkpointing=False
```

* revert NF4 changes from https://github.com/pytorch/ao/pull/360
* FSDP2 + QLoRA multi-gpu test: `pytest -s test/dtypes/test_nf4.py -k test_qlora`: e2e fsdp2 + qlora test
* NF4.clone test `pytest -s test/dtypes/test_nf4.py -k test_tensor_copy`: torchtune implemented `NF4.clone`, upstream it to TorchAO. This is needed by unit test `copy.deepcopy(model)`

### CODE REVIEW

File: torchao/dtypes/nf4tensor.py [hunk 1]
```diff
@@ -11,10 +11,6 @@
 from torch import Tensor
 from torch.distributed.device_mesh import DeviceMesh
 from torch._prims_common import make_contiguous_strides_for
-from torchao.dtypes.utils import (
```
Comment (weifengpy) [inline torchao/dtypes/nf4tensor.py:14]: https://github.com/pytorch/ao/pull/360 consolidate `_implements` and `_ATEN_OP_OR_TORCH_FN_TABLE` but it breaks torchtune. revert for now to unblock torchtune quickly

File: test/dtypes/test_nf4.py [hunk 2]
```diff
@@ -431,6 +439,170 @@ def test_to_cpu(self):
...
+
+    @pytest.mark.skipif(
+        version.parse(torch.__version__).base_version < "2.4.0",
+        reason="torch >= 2.4 required",
+    )
+    @skip_if_lt_x_gpu(2)
+    def test_qlora_fsdp2(self):
+        from torch.distributed._composable.fsdp import CPUOffloadPolicy, OffloadPolicy
+
+        self.run_subtests(
```
Comment (weifengpy) [inline test/dtypes/test_nf4.py:497]: e2e mutli-gpu FSDP + QLoRA test should be able to catch regression in the future

File: torchao/dtypes/nf4tensor.py (hunk 1 above)
Comment (drisspg) [reply torchao/dtypes/nf4tensor.py:14]: Do you know how exactly this breaks torchtune? is it a versioning thing between saved models and this new model?

Comment (weifengpy) [reply torchao/dtypes/nf4tensor.py:14]: the error is `TypeError: nf4_detach() missing 1 required positional argument: 'args'`. So there is something incompatiable around with NF4 around `_ATEN_OP_OR_TORCH_FN_TABLE[func]`

the errors shows up when people start training in TorchTune for the 1st time

Comment (drisspg) [reply torchao/dtypes/nf4tensor.py:14]: @jerryzh168  any thoughts on why this is happening, otherwise are you okay to undo your changes?

Comment (ebsmothers) [reply torchao/dtypes/nf4tensor.py:14]: @drisspg to add a bit more to what @weifengpy already said, the full stack trace is [here](https://gist.github.com/ebsmothers/154bafd723d6d58f3d61a3b42e3e913f).

Comment (drisspg) [reply torchao/dtypes/nf4tensor.py:14]: Screenshot 2024-06-21 at 11 22 11 AM


I think there is something weird going on with the torch_function dispatch, where the args are becoming the arg name for aten.. If someone wants to track down how this should work and why this is I am for it but otherwhise I will approve to unblock

**Review by drisspg — APPROVED** [review]

**Pull request merged**